src/coprocessor/codec/data_type/vector.rs

```diff
@@ -443,6 +443,160 @@ impl VectorValue {
     }
+
+    /// Returns maximum encoded size in binary format.
+    pub fn maximum_encoded_size(&self) -> Result<usize> {
+        match self {
+            VectorValue::Int(ref vec) => Ok(vec.len() * 9),
+
+            // Some elements might be NULLs which encoded size is 1 byte. However it's fine because
+            // this function only calculates a maximum encoded size (for constructing buffers), not
+            // actual encoded size.
+            VectorValue::Real(ref vec) => Ok(vec.len() * 9),
+            VectorValue::Decimal(ref vec) => {
+                let mut size = 0;
+                for el in vec {
+                    match el {
+                        Some(v) => {
+                            // FIXME: We don't need approximate size. Maximum size is enough (so
+                            // that we don't need to iterate each value).
+                            size += 1 /* FLAG */ + v.approximate_encoded_size();
+                        }
+                        None => {
+                            size += 1;
+                        }
+                    }
+                }
+                Ok(size)
+            }
+            VectorValue::Bytes(ref vec) => {
+                let mut size = 0;
+                for el in vec {
+                    match el {
+                        Some(v) => {
+                            size += 1 /* FLAG */ + 10 /* MAX VARINT LEN */ + v.len();
+                        }
+                        None => {
+                            size += 1;
+                        }
+                    }
+                }
+                Ok(size)
+            }
+            VectorValue::DateTime(ref vec) => Ok(vec.len() * 9),
+            VectorValue::Duration(ref vec) => Ok(vec.len() * 9),
+            VectorValue::Json(ref vec) => {
+                let mut size = 0;
+                for el in vec {
+                    match el {
+                        Some(v) => {
+                            size += 1 /* FLAG */ + v.binary_len();
+                        }
+                        None => {
+                            size += 1;
+                        }
+                    }
+                }
+                Ok(size)
+            }
+        }
+    }
+
+    /// Encodes a single element into binary format.
+    pub fn encode(&self, row_index: usize, output: &mut Vec<u8>) -> Result<()> {
+        use crate::coprocessor::codec::mysql::DecimalEncoder;
+        use crate::coprocessor::codec::mysql::JsonEncoder;
+        use crate::util::codec::bytes::BytesEncoder;
+        use crate::util::codec::number::NumberEncoder;
+
+        match self {
+            VectorValue::Int(ref vec) => {
+                match vec[row_index] {
+                    None => {
+                        output.push(datum::NIL_FLAG);
+                    }
+                    Some(val) => {
+                        output.push(datum::INT_FLAG);
+                        output.encode_i64(val)?;
```

Comment:
It seems now in the `master` we will comparable `i64` into `var_i64`.

Reply:
Yes. It utilizes the same feature as above that (looks like) TiDB can correctly parse datum based on the flag in the datum instead of the given field type. (@XuHuaiyu Please help confirm)

The reason that use `i64` instead of `var_i64` is to be efficient.

Reply:
> The reason that uses `i64` instead of `var_i64` is to be efficient.

What's the result of the comparison? This may reduce the transmission efficiency in most cases.

Reply:
See https://github.com/tikv/tikv/pull/3629

Reply:
How about the benchmark for YCSB? It seems there is a story why we use `varint ` to short the value in network transmission @coocood

Reply:
@AndreMouche As long as you switch to Arrow format later, it will be the same. In Arrow we cost even more on network. I think the main concern here should be whether or not TiDB can correctly accept this behaviour.